wordpress-mobile/WordPress-iOS
# [Gutenberg] Set consistent editor colours in both the WordPress and Jetpack apps

Fixes the iOS side of https://github.com/wordpress-mobile/gutenberg-mobile/issues/5047, [Android PR is here](https://github.com/wordpress-mobile/WordPress-Android/pull/16968/)

## Description

In the Jetpack app, some UI elements like the cursor, text selection, and caret are using a green colour. However, the rest of the editor’s elements are using blue. 

To be consistent with colours across the editor, this PR seeks to update them for consistency across both the WordPress and Jetpack apps. The PR specifically achieves the following for the Jetpack app:

* Updates the colour of UI elements within the main editor (e.g. cursor, highlight) from green to blue.
* Updates the accent of sub screens within Post Settings from green to blue.

There's still some known inconsistencies that need to be addressed in separate PRs, however:

* The media picker (available from media blocks and the Post Settings screen) still has some green.
* There is still green visible in other screens 'connected' to the editor, such as Preview, History, Help & Support, etc.

These remaining inconsistencies have been captured in the issue at https://github.com/wordpress-mobile/gutenberg-mobile/issues/4889 for the future.

## Screenshots

A before/after screenshot taken from the Jetpack app to show the editor's green → blue colour changes:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/180868507-303c44fb-23b4-474b-93fa-2c18d299f317.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/180877183-cec91c87-798f-42ab-b376-b1bd53e20640.png" width="100%"> |

A before/after taken from the WordPress app (dark mode) to confirm there are no unexpected changes:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/181541724-efb97952-004d-4974-b774-a6d7968b5710.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/181542827-627e6555-0d0f-4109-8876-c1fe1d9cbadf.png" width="100%"> |

## Testing

* Download this PR's installable build for both the WordPress and Jetpack apps.
* In each app, navigate to the editor.
* Verify the colours used throughout the editor appear consistent for both app. The cursor, text selection, and caret should all now be blue for Jetpack (this is in contrast to the green that's used outside of the app's editor). There should be no change in the WordPress app, with the colours remaining blue.  

Special considerations to keep in mind while testing include:

* This PR aims to bring colour consistency to the Post Settings screen as well as the editor.
* Testing of any popovers, modals, or less visible UI elements within the editor would be beneficial. 
* As this is a UI change, dark/light mode should be a consideration.
* These changes should be noticeable in both the post _and_ the page editor.

## Code Changes

The colours we're aiming to change are using the default `.tintColor`, which is set to `.primary` via the following line of code:

https://github.com/wordpress-mobile/WordPress-iOS/blob/3eabde78c08cd210dfe1ed26109279d32d448ad7/WordPress/Classes/Extensions/Colors%20and%20Styles/WPStyleGuide%2BApplicationStyles.swift#L20-L22

`.primary` is set as `.jetpackGreen` [here for the Jetpack app](https://github.com/wordpress-mobile/WordPress-iOS/blob/a96da1dbcf200eb532e83666fb12fca5c1812129/WordPress/Jetpack/AppStyleGuide.swift#L19) and `.blue` [here for the WordPress app](https://github.com/wordpress-mobile/WordPress-iOS/blob/a96da1dbcf200eb532e83666fb12fca5c1812129/WordPress/Classes/Utility/App%20Configuration/AppStyleGuide.swift#L18).

As a number of UI elements needed to be updated as part of this PR, a new `.editorPrimary` variable was created, following the same patterns found for other colours in [the `UIColor+MurielColors.swift` file](https://github.com/wordpress-mobile/WordPress-iOS/blob/d6b813b903c34d4d66ac07c5e452b378fd952e63/WordPress/Classes/Extensions/Colors%20and%20Styles/UIColor+MurielColors.swift), such as `.primary` and `.success`.

### Root React Native View

The `tintColor` for the editor's main view (i.e. the root React Native view) has been set here: 

https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L568

### React Native Modals

`configureDefaultTint()` is [setting the tint colour for the entire window](https://github.com/wordpress-mobile/WordPress-iOS/blob/3eabde78c08cd210dfe1ed26109279d32d448ad7/WordPress/Classes/Extensions/Colors%20and%20Styles/WPStyleGuide%2BApplicationStyles.swift#L20-L22) of each screen via [`UIWindow`](https://developer.apple.com/documentation/uikit/uiwindow). As React Native's modals exist outside of the root view, it's necessary to override the default `tintColor` once again to ensure consistency in `BottomSheets` and other parts of the app that may be using modals.

This was achieved on the native side via `override func present`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L415-L420

### React Native Action Sheets and Switches

The default `.tintColor` for switches/toggles within the editor's modals needed to be overridden, which was done via the following line of code:

https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L576

Note, the above was set within `GutenbergViewController.swift` rather than `WordPressAppDelegate.swift` as `RCTModalHostViewController` resulted in an out-of-scope error in the latter's file.

In addition, the `.tintColor` for action sheets has been set directly in React Native via https://github.com/wordpress-mobile/gutenberg-mobile/pull/5074.

### Post Settings Screen

It was possible to set the `tintColor` for multiple UI elements in the Post Settings screen.

Some hacky code has been used for the parent category screen (found within Post Settings). This was necessary because the screen is used elsewhere in the app (via Site Settings → Category) and there wasn't any other views in the hierarchy that could be used to only target the Post Settings UI:

https://github.com/wordpress-mobile/WordPress-iOS/blob/56dab966a361844f1e30cdea8ce4d71218d0ae1b/WordPress/Classes/System/WordPressAppDelegate.swift#L889-L893


## Regression Notes
1. Potential unintended areas of impact

There's the potential for this PR to unexpectedly change colours in either the WordPress or Jetpack app beyond what's desired. For example, the cursor colour may unexpectedly change within the WordPress app. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I manually tested both apps, making sure to experiment with various blocks and writing flows.

3. What automated tests I added (or what prevented me from doing so)

No automated tests.

<hr />

PR submission checklist:
